coral-spark/src/main/java/com/linkedin/coral/spark/CoralSpark.java

```diff
@@ -115,6 +118,43 @@ private static String constructSparkSQL(RelNode sparkRelNode) {
     return rewrittenSparkSqlNode.toSqlString(SparkSqlDialect.INSTANCE).getSql();
   }
 
+  /**
+   * Users use this function as the main API for getting CoralSpark instance.
+   *
+   * Internally Appropriate parts of Sql RelNode is converted to Spark RelNode, Spark RelNode is converted back
+   * to SqlNode and SqlNode to SparkSQL.
+   *
+   * It returns an instance of CoralSpark which contains
+   *  1) Spark SQL
+   *  2) Base tables
+   *  3) Spark UDF information objects, ie. List of {@link SparkUDFInfo}
+   *
+   * @param sqlNode CoralNode which will be translated to SparkSQL.
+   * @param convertor Functional Interface to convert SqlNode to appropriate RelNode
+   *
+   *
+   * @return [[CoralSparkInfo]]
+   */
+  public static CoralSpark create(SqlNode sqlNode, Function<SqlNode, RelNode> convertor){
```

Comment:
Note that we are assuming this is the `Coral IR SqlNode`. Assuming that `Coral IR SqlNode` has array indexes start from 1 (see `standardizeRel()`, I am pretty sure the array index in this tree will start from 0. For consistency, somehow we need a `SELECT` node here that:
1-  is a`SqlNode`
2- has array indexes starting from 1.
I think this implies that the caller to this method must have taken the query already, converted the `SELECT SqlNode` to `RelNode` then back to `SELECT SqlNode`. The former `SELECT` will have array indexes start from 0, and the latter will have them start from 1. What remains after the caller has done this step is that the latter `Coral IR SqlNode` is simply converted to Spark SQL string (using the Spark dialect and after applying the`SparkSqlRewriter` (on the latter `SqlNode`), without having to go through `RelNode` here.
The caller can call the `RelNode` based `create` on the obtained `RelNode` in the caller. It is best to code this logic in a test case. @aastha25 @ljfgem FYI.

Reply:
Yes, currently SqlNode given as input to this function has `0-based indexing. Now as you have mentioned that we need a `SELECT`node here should have array indexes starting from 1, so it means all `RelNode` based transformations, would have been done by the caller of this function. It means the caller should be able to execute the below transformations on the `SELECT `node before calling this new create function.
```
    if(sqlNode instanceof SqlCommand) {
      SqlNode selectNode = ((SqlCreateTable) sqlNode).getSelectQuery();
      sparkRelInfo = IRRelToSparkRelTransformer.transform(convertor.apply(selectNode));
      selectNode = new CoralRelToSqlNodeConverter().convert(sparkRelInfo.getSparkRelNode());
      ((SqlCreateTable) sqlNode).setSelectQuery(selectNode);
    } else {
      sparkRelInfo = IRRelToSparkRelTransformer.transform(convertor.apply(sqlNode));
      sqlNode = new CoralRelToSqlNodeConverter().convert(sparkRelInfo.getSparkRelNode());
    }
```
 Should we create a separate function in `CoralSpark` to handle this `SqlNode`->`RelNode`->`SqlNode` transformation? 
 I don't think, we can call `RelNode`-based create as currently, it will convert `RelNode` to `SQL` directly.

Reply:
in the translation path:
`Language1 (hiveSQL) --> CoralSqlNode (LHS) --> CoralRelNode --> CoralSqlNode (RHS) --> Language 2 (spark SQL) `

The `coral-spark` class is responsible for generating sparkSQL from CoralIR. It currently accepts the `CoralRelNode` and applies the RHS side of transformations to generate a spark-compatible SQL. In your current implementation of `create(..)`, you are passing CoralSqlNode (LHS) and using `convertor.apply(selectNode)` to generate `CoralRelNode`. Could you please refactor to apply the LHS side of transformations outside this class and then invoke existing method `create(RelNode irRelNode){..}` to generate the updated `selectNode`?

Reply:
@nimesh1601 could you summarize the offline discussion/agreement here?